src/Security/PasswordExpirationMiddleware.php

```diff
@@ -0,0 +1,328 @@
+ *   - else if the change password form URL is set, then redirect to it
+ *   - else set current user to null (deauthenticate for the current request) and process further
+ */
+class PasswordExpirationMiddleware implements HTTPMiddleware
+{
+    use Configurable;
+
+    /**
+     * Session key for persisting URL of the password change form
+     */
+    const SESSION_KEY_REDIRECT = __CLASS__.'.change password redirect';
+
+    /**
+     * Session key for persisting Array of urls allowed to visit by users
+     * with passwords expired (apart from the password change form)
+     */
+    const SESSION_KEY_WHITELIST = __CLASS__.'.whitelisted urls';
+
+    /**
+     * Session key for persisting a flag allowing to process the current request
+     * without performing password expiration check
+     */
+    const SESSION_KEY_ALLOW_CURRENT_REQUEST = __CLASS__.'.allow current request';
+
+    /**
+     * Enable the middleware to check password expiration
+     *
+     * @var bool
+     *
+     * @config
+     */
+    private static $enable = true;
+
+    /**
+     * List of URLs allowed to visit by users with password expired
+     * These URLs won't be redirected to the password change form.
+     * Requests to these URLs will have the Member NOT deauthenticated
+     * in case there is no password change form registered.
+     *
+     * @var array<string>
+     *
+     * @config
+     */
+    private static $whitelisted_urls = [];
+
+    /**
+     * List of URL patterns allowed for users to visit where
+     * URL starts with the pattern
+     *
+     * @var array<string>
+     *
+     * @config
+     */
+    private static $whitelisted_url_startswith = [];
+
+    /**
+     * Where users with expired passwords get redirected by default
+     * when login form didn't register a custom one with
+     * {@see SilverStripe\Security\AuthenticationMiddleware::setRedirect}
+     *
+     * @var string
+     *
+     * @config
+     */
+    private static $default_redirect = null;
+
+    /**
+     * Whether we should allow the current request to be finished as is even if
+     * the user password has been expired.
+     *
+     * WARNING: this is not a configurable attribute, but should be controlled by a LoginHandler
+     *
+     * @var bool
+     */
+    private static $allowCurrentRequest = false;
+
+    /**
+     * The list of mimetypes allowing a redirect to a change password form.
+     * By default this is (x)HTML
+     *
+     * @var array<string>
+     *
+     * @config
+     */
+    private static $mimetypes_allowing_redirect = [
+        '*/*',
+        'text/*',
+        'text/html',
+        'application/xhtml+xml',
+        'text/xml',
+        'application/xml'
+    ];
+
+    /**
+     * Identify the current user from the request
+     *
+     * @param HTTPRequest $request
+     * @param callable $delegate
+     * @return HTTPResponse
+     */
+    public function process(HTTPRequest $request, callable $delegate)
```

Comment:
You _could_ add an HTTPResponse type hint here. If you do your tests will fail, because the delegates you have in the tests don't return HTTPResponses

Reply:
that's the definition from the `HTTPMiddleware` interface. I'll just remove the doc

Reply:
Yeah but you can add return types in subclasses if you want to, it was just a suggestion - keep the doc block 😄

Reply:
Yeah, as you mentioned yourself, we cannot guarantee everything actually returns HTTPResponse, so that may actually be a breaking change :)

Reply:
lol - touche